File: DongbinBook/sorting/quick_sort.py
arr = [7, 5, 9, 0, 3, 1, 6, 2, 4, 8]


def quick_sort(arr, start, end):
    if start >= end:
        return
    pivot = start
    left = start + 1
    right = end
    while left <= right:
        # pivot 보다 큰 데이터 찾을 때까지 반복
        while left <= end and arr[left] <= arr[pivot]:
            left += 1
        # pivot 보다 작은 데이터 찾을 때까지 반복
        while right > start and arr[right] >= arr[pivot]:
            right -= 1
        # 엇갈린 경우
        if left > right:
            arr[right], arr[pivot] = arr[pivot], arr[right]
        else:
            # 작은 데이터 큰데이터 교환
            arr[left], arr[right] = arr[right], arr[left]
    # 분할 이후 왼쪽 부분과 오른쪽 부분에서 각각 정렬 수행
    quick_sort(arr, start, right - 1)
    quick_sort(arr, right + 1, end)


def quick_sort_pythonic(arr):
    if len(arr) <= 1:
        return arr

    pivot = arr[0]
    tail = arr[1:]

    left_side = [x for x in tail if x <= pivot]  # 분할된 왼쪽부분
    right_side = [x for x in tail if x > pivot]  # 분할된 오른쪽부분

    return quick_sort_pythonic(left_side) + [pivot] + quick_sort_pythonic(right_side)


print(quick_sort_pythonic(arr))

File: Java/src/com/dobi/algo/algobook1/sort/QuickSort.java
package com.dobi.algo.algobook1.sort;

public class QuickSort {
    public static void quickSort(int[] arr, int start, int end) {
        if (start >= end) return; // 원소가 1개인 경우 종료
        int pivot = start;
        int left = start + 1;
        int right = end;

        while (left <= right) {
            // pivot 보다 큰 데이터 찾을 때 까지 반복
            while (left <= end && arr[left] <= arr[pivot]) left++;
            // pivot 보다 작은 데이터 찾을 때 까지 반복
            while (right > start && arr[right] >= arr[pivot]) right--;

            // 엇갈린 경우 작은 데이터와 피벗 교체
            if (left > right) {
                int temp = arr[pivot];
                arr[pivot] = arr[right];
                arr[right] = temp;
            }
            // 엇갈리지 않은 경우 작은 데이터와 큰 데이터 교체
            else {
                int temp = arr[left];
                arr[left] = arr[right];
                arr[right] = temp;
            }
        }
        quickSort(arr, start, right-1);
        quickSort(arr, right+1, end);
    }
    public static void main(String[] args) {
        int n = 10;
        int[] arr = {7,5,3,1,4,10,9,8, 11,12};

        quickSort(arr, 0, n-1);

        for (int i = 0; i < n; i++) {
            System.out.println(arr[i]);
        }
    }
}
